src/meta/meta_ckpt.c

```diff
@@ -503,42 +538,213 @@ __wt_meta_blk_mods_load(WT_SESSION_IMPL *session, const char *config, WT_CKPT *c
  */
 int
 __wt_meta_ckptlist_get(
-  WT_SESSION_IMPL *session, const char *fname, bool update, WT_CKPT **ckptbasep)
+  WT_SESSION_IMPL *session, const char *fname, bool update, WT_CKPT **ckptbasep, size_t *allocated)
 {
     WT_DECL_RET;
     char *config;
 
     config = NULL;
 
     WT_ERR(__wt_metadata_search(session, fname, &config));
-    WT_ERR(__wt_meta_ckptlist_get_from_config(session, update, ckptbasep, config));
+    WT_ERR(__wt_meta_ckptlist_get_from_config(session, update, ckptbasep, allocated, config));
 
 err:
     __wt_free(session, config);
     return (ret);
 }
 
+#ifdef HAVE_DIAGNOSTIC
+/*
+ * __assert_ckpt_matches --
+ *     Assert that given two checkpoints match.
+ */
+static void
+__assert_ckpt_matches(WT_SESSION_IMPL *session, WT_CKPT *ckpt_a, WT_CKPT *ckpt_b)
+{
+    /*
+     * We are not checking checkpoint time, because there could be a minute difference depending
+     * upon when the checkpoint information was generated. This is acceptable.
+     */
+    WT_ASSERT(session,
+      (ckpt_a->name == NULL && ckpt_b->name == NULL) ||
+        (ckpt_a->name != NULL && ckpt_b->name != NULL && strcmp(ckpt_a->name, ckpt_b->name) == 0));
+    WT_ASSERT(session, ckpt_a->order == ckpt_b->order);
+    WT_ASSERT(session, ckpt_a->size == ckpt_b->size);
+    WT_ASSERT(session, ckpt_a->write_gen == ckpt_b->write_gen);
+    WT_ASSERT(session, ckpt_a->run_write_gen == ckpt_b->run_write_gen);
+    WT_ASSERT(session,
+      ckpt_a->ta.newest_start_durable_ts == ckpt_b->ta.newest_start_durable_ts &&
+        ckpt_a->ta.newest_stop_durable_ts == ckpt_b->ta.newest_stop_durable_ts &&
+        ckpt_a->ta.oldest_start_ts == ckpt_b->ta.oldest_start_ts &&
+        ckpt_a->ta.newest_txn == ckpt_b->ta.newest_txn &&
+        ckpt_a->ta.newest_stop_ts == ckpt_b->ta.newest_stop_ts &&
+        ckpt_a->ta.newest_stop_txn == ckpt_b->ta.newest_stop_txn &&
+        ckpt_a->ta.prepare == ckpt_b->ta.prepare &&
+        __wt_check_addr_validity(session, &ckpt_a->ta, true) == 0);
+    // WT_ASSERT(session, compare ckpt_a->addr.data, ckpt_b->addr.data, ckpt_a->addr.size == 0);
```

Comment:
I noticed that the checkpoint addr mismatches between the cached checkpoint and the read checkpoint. I am wondering what could be the reason for that. It is possible the addr for cached checkpoints do not matter, hence the testing doesn't show this to be a problem.

Need to find an answer before getting rid of the assert.

Reply:
Offhand I do not know the answer to that. Is this caching the checkpoint information before the addr string is finalized?

Reply:
I would not expect the addresses to be different -- when you decode the differing addresses, what specific difference do you see?

Reply:
I re-added the assert locally, and for instance `ex_backup` fails
`WT_ASSERT(session, memcmp(ckpt_a->addr.data, ckpt_b->addr.data, ckpt_a->addr.size) == 0);`

```
(gdb) p ckpt_a->addr
$1 = {data = 0x954720, size = 73, mem = 0x954720, memsize = 73, flags = 0}
(gdb) p ckpt_b->addr
$2 = {data = 0x952350, size = 72, mem = 0x952350, memsize = 72, flags = 0}
(gdb) p *ckpt_a
$3 = {name = 0x637320 "WiredTigerCheckpoint.1", order = 1, sec = 1620618278, size = 266240, write_gen = 11, run_write_gen = 1,
  block_metadata = 0x94ee90 "encryption=(keyid=,name=),block_metadata_encrypted=false,block_metadata=[access_pattern_hint=none,allocation_size=4KB,app_metadata=,assert=(commit_timestamp=none,durable_timestamp=none,read_timestamp="...,
  block_checkpoint = 0x7fbe70 "checkpoint=(WiredTigerCheckpoint.1=(addr=\"01c00081e456dcf599c00181e46c140199808080808080cfc0e303dfc0\",order=1,time=1620618278,size=266240,newest_start_durable_ts=0,oldest_start_ts=0,newest_txn=0,newes"..., backup_blocks = {{id_str = 0x0,
      bitstring = {data = 0x0, size = 0, mem = 0x0, memsize = 0, flags = 0}, nbits = 0, offset = 0, granularity = 0, flags = 0}, {id_str = 0x0, bitstring = {data = 0x0, size = 0, mem = 0x0, memsize = 0, flags = 0}, nbits = 0, offset = 0, granularity = 0, flags = 0}},
  ta = {newest_start_durable_ts = 0, newest_stop_durable_ts = 0, oldest_start_ts = 0, newest_txn = 0, newest_stop_ts = 18446744073709551615, newest_stop_txn = 18446744073709551605, prepare = 0 '\000', init_merge = 1 '\001'}, addr = {data = 0x954720, size = 73,
    mem = 0x954720, memsize = 73, flags = 0}, raw = {data = 0x84f1b0, size = 36, mem = 0x84f1b0, memsize = 127, flags = 0}, bpriv = 0x0, flags = 0}
(gdb) p *ckpt_b
$4 = {name = 0x84f120 "WiredTigerCheckpoint.1", order = 1, sec = 1620618278, size = 266240, write_gen = 11, run_write_gen = 1, block_metadata = 0x0, block_checkpoint = 0x0, backup_blocks = {{id_str = 0x0, bitstring = {data = 0x0, size = 0, mem = 0x0, memsize = 0,
        flags = 0}, nbits = 0, offset = 0, granularity = 0, flags = 0}, {id_str = 0x0, bitstring = {data = 0x0, size = 0, mem = 0x0, memsize = 0, flags = 0}, nbits = 0, offset = 0, granularity = 0, flags = 0}}, ta = {newest_start_durable_ts = 0, 
    newest_stop_durable_ts = 0, oldest_start_ts = 0, newest_txn = 0, newest_stop_ts = 18446744073709551615, newest_stop_txn = 18446744073709551605, prepare = 0 '\000', init_merge = 0 '\000'}, addr = {data = 0x952350, size = 72, mem = 0x952350, memsize = 72, flags = 0},
  raw = {data = 0x965d30, size = 36, mem = 0x965d30, memsize = 36, flags = 0}, bpriv = 0x0, flags = 0}
```

The size of the address is different itself, off by one. The same issue with ex_backup_block. There are several python tests that fail as well, I checked couple of them, and the addr size is off by 1.

Here is another example from `test_bug010.py`

```
(gdb) p ckpt_a->addr
$3 = {data = 0x1c9d560, size = 59, mem = 0x1c9d560, memsize = 59, flags = 0}
(gdb) p ckpt_b->addr
$4 = {data = 0x7f973c008fa0, size = 58, mem = 0x7f973c008fa0, memsize = 58, flags = 0}
(gdb) x /59x 0x1c9d560
0x1c9d560:      0x31383130      0x34653138      0x38613562      0x31306434
0x1c9d570:      0x31383238      0x35313465      0x64623634      0x33383631
0x1c9d580:      0x33653138      0x62633032      0x30386433      0x30383038
0x1c9d590:      0x66323265      0x66633063      0x00003063      0x00000000
0x1c9d5a0:      0xffffffff      0xffffffff      0x00000401      0x00000000
0x1c9d5b0:      0x01c9ad90      0x00000000      0x01c9e2e0      0x00000000
0x1c9d5c0:      0x00000000      0x00000000      0x00000000      0x00000000
0x1c9d5d0:      0x5f6e6f69      0x657a6973      0x424b343d      0x7070612c
0x1c9d5e0:      0x74656d5f      0x74616461      0x612c3d61      0x72657373
0x1c9d5f0:      0x63283d74      0x696d6d6f      0x69745f74      0x7473656d
0x1c9d600:      0x3d706d61      0x656e6f6e      0x7275642c      0x656c6261
0x1c9d610:      0x6d69745f      0x61747365      0x6e3d706d      0x2c656e6f
0x1c9d620:      0x64616572      0x6d69745f      0x61747365      0x6e3d706d
0x1c9d630:      0x2c656e6f      0x74697277      0x69745f65      0x7473656d
0x1c9d640:      0x3d706d61      0x2966666f      0x6f6c622c
(gdb) x /58x 0x7f973c008fa0
0x7f973c008fa0: 0x31383130      0x34653138      0x38613562      0x31306434
0x7f973c008fb0: 0x31383238      0x35313465      0x64623634      0x33383631
0x7f973c008fc0: 0x33653138      0x62633032      0x30386433      0x30383038
0x7f973c008fd0: 0x66323265      0x66633063      0x30633063      0x00000000
0x7f973c008fe0: 0x00000000      0x00000000      0x00000035      0x00000000
0x7f973c008ff0: 0xe4818101      0x014da8b5      0x15e48182      0x8316bd46
0x7f973c009000: 0xcb20e381      0x8080803d      0xcfc02fe2      0x0000c0c0
0x7f973c009010: 0x00000000      0x00000000      0x00000065      0x00000000
0x7f973c009020: 0x3c009130      0x00007f97      0x3c009160      0x00007f97
0x7f973c009030: 0x00000000      0x00000000      0x00000000      0x00000000
0x7f973c009040: 0x00000000      0x00000000      0x00000000      0x00000000
0x7f973c009050: 0x00000000      0x00000000      0x00000000      0x00000000
0x7f973c009060: 0x00000000      0x00000000      0x00000000      0x00000000
0x7f973c009070: 0x00000000      0x00000000      0x000000b5      0x00000000
0x7f973c009080: 0x53ada7a0      0x00007f97
```

The initial portion of the memory dump matches between the two checkpoints. First 16 bytes? Unfortunately, I do not have the expertise in figuring this one out, it can take me a while. @keithbostic , is it obvious to you what is happening here ?